src/connector/src/sink/file_sink/mod.rs

```diff
@@ -0,0 +1,230 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+pub mod fs;
+pub mod gcs;
+pub mod opendal_sink;
+pub mod s3;
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use anyhow::anyhow;
+use arrow_schema::{DataType as ArrowDataType, Field as ArrowField, SchemaRef};
+use async_trait::async_trait;
+use opendal::{Operator, Writer as OpendalWriter};
+use parquet::arrow::AsyncArrowWriter;
+use parquet::file::properties::WriterProperties;
+use risingwave_common::array::{to_record_batch_with_schema, Op, StreamChunk};
+use risingwave_common::buffer::Bitmap;
+use risingwave_common::catalog::Schema;
+
+use crate::sink::catalog::SinkEncode;
+use crate::sink::{Result, SinkError, SinkWriter};
+
+// Todo(wcy-fdu): upgrade to opendal 0.47.
+const SINK_WRITE_BUFFER_SIZE: usize = 16 * 1024 * 1024;
+
+pub struct OpenDalSinkWriter {
+    schema: SchemaRef,
+    operator: Operator,
+    sink_writer: Option<FileWriterEnum>,
+    pk_indices: Vec<usize>,
+    is_append_only: bool,
+    write_path: String,
+    epoch: Option<u64>,
+    executor_id: u64,
+    encode_type: SinkEncode,
+}
+
+/// The `FileWriterEnum` enum represents different types of file writers used for various sink
+/// implementations.
+///
+/// # Variants
+///
+/// - `ParquetFileWriter`: Represents a Parquet file writer using the `AsyncArrowWriter<W>`
+/// for writing data to a Parquet file. It accepts an implementation of W: `AsyncWrite` + `Unpin` + `Send`
+/// as the underlying writer. In this case, the `OpendalWriter` serves as the underlying writer.
+///
+/// - `FileWriter`: Represents a file writer for sinks other than Parquet. It uses the `OpendalWriter`
+/// directly for writing data to the file.
+///
+/// The choice of writer used during the actual writing process depends on the encode type of the sink.
+enum FileWriterEnum {
+    ParquetFileWriter(AsyncArrowWriter<OpendalWriter>),
+    FileWriter(OpendalWriter),
+}
+
+#[async_trait]
+impl SinkWriter for OpenDalSinkWriter {
+    async fn write_batch(&mut self, chunk: StreamChunk) -> Result<()> {
+        // Note: epoch is used to name the output files.
+        // Todo: after enabling sink decouple, use the new naming convention.
+        let epoch = self.epoch.ok_or_else(|| {
+            SinkError::Opendal("epoch has not been initialize, call `begin_epoch`".to_string())
+        })?;
+        if self.sink_writer.is_none() {
+            self.create_sink_writer(epoch).await?;
+        }
+        if self.is_append_only {
+            self.append_only(chunk).await
+        } else {
+            // currently file sink only supports append only mode.
+            unimplemented!()
```

Comment:
Is here reachable or not? In other words, do we have logics somewhere to ensure user will get an error instead of panicing the CN if the user ingests non-append-only data to file sink? We should probably implement the [`validate`](https://github.com/risingwavelabs/risingwave/blob/ac20951e584ff149ea494f06a8dc68810f8e3547/src/connector/src/sink/file_sink/opendal_sink.rs#L79) method accordingly instead of leaving it empty.

Reply:
If Any of these two (`type = 'append-only', `force_append_only='true'`) is not specified in the create sink statement.
will get error in frontend:
```
Caused by these errors (recent errors listed first):
  1: Sink error
  2: The sink cannot be append-only. Please add "force_append_only='true'" in FORMAT ENCODE options to force the sink to be append-only. Notice that this will cause the sink executor to drop any UPDATE or DELETE message.
```